os/arch/arm/src/amebad/amebad_reboot_reason.c

```diff
@@ -65,36 +65,44 @@
 
 static reboot_reason_code_t backup_reg;
 
-void up_reboot_reason_init(void)
-{
-	/* Read the same backup register for the boot reason */
-	backup_reg = BKUP_Read(BKUP_REG1);
-	BKUP_Write(BKUP_REG1, REBOOT_REASON_INITIALIZED);
-
-}
-
-reboot_reason_code_t up_reboot_reason_read(void)
+static void up_reboot_reason_get_hw_value(void)
 {
 	u32 boot_reason = 0;
 
 	if (backup_reg != REBOOT_REASON_INITIALIZED) {
-		return backup_reg;
+		return;
 	} else {
 		/* Read AmebaD Boot Reason, WDT and HW reset supported */
 		boot_reason = BOOT_Reason();
 
 		/* HW reset */
 		if (boot_reason == 0) {
-			return REBOOT_SYSTEM_HW_RESET;
+			backup_reg = REBOOT_SYSTEM_HW_RESET;
+			return;
 		}
 
 		/* KM4 or KM0 WDT reset */
 		else if ((boot_reason & BIT_BOOT_KM4WDG_RESET_HAPPEN) || (boot_reason & BIT_BOOT_WDG_RESET_HAPPEN)) {
-			return REBOOT_SYSTEM_WATCHDOG;
+			backup_reg = REBOOT_SYSTEM_WATCHDOG;
+			return;
 		}
 	}
 
-	return REBOOT_UNKNOWN;
+	backup_reg = REBOOT_UNKNOWN;
+}
+
+reboot_reason_code_t up_reboot_reason_init(void)
+{
+	/* Read the same backup register for the boot reason */
+	backup_reg = BKUP_Read(BKUP_REG1);
+	up_reboot_reason_get_hw_value();
+	BKUP_Write(BKUP_REG1, REBOOT_REASON_INITIALIZED);
+	return backup_reg;
```

Comment:
It looks strange that the function returns backup_reg without any code controlled. How about getting return value from get_hw_val and saving it in this and then return?